### PR TITLE
Split sync and async API

### LINKED ISSUE #47: macOS: Notes on notification detection
The [current detection mechanism](https://github.com/frewsxcv/rust-dark-light/blob/925fe9dd7cce2701fb9f5958260cd75c94cdf552/src/platforms/macos/notify.rs) just runs `detect` in a loop, which is woefully inefficient. Details on more efficient implementation options in https://github.com/frewsxcv/rust-dark-light/pull/37#issuecomment-2123412904 and https://github.com/cormacrelf/dark-notify/blob/master/src/app.rs.

Also, it doesn't actually work, see https://github.com/frewsxcv/rust-dark-light/issues/29. I suspect that the problem is that we need to be running the main application with `NSApplication::run` or `NSApplicationMain` for this to work? But semantically, that probably isn't really something that `rust-dark-light` should be doing, so unsure what to do.

Even if that isn't the problem, or it can be worked around by using `"AppleInterfaceThemeChangedNotification"`, another problem is that most Rust async runtimes don't support mach ports, and as such wouldn't update since notifications arrive via. those, I think? See https://github.com/madsmtm/objc2/issues/279 for that.

### PR DESCRIPTION
**Pull request by edfloreshz**
This pull request introduces several changes to the codebase to add support for both synchronous and asynchronous detection and subscription to system theme changes. 

The changes include modifications to dependencies, new feature flags, and refactoring of existing code to accommodate the new functionality.

> [!WARNING]  
> These are breaking changes, we should consider shipping v2.0.0 after merging this.

Dependency and feature flag updates:

* [`Cargo.toml`](diffhunk://#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R12-R19): Added `sync` feature flag and updated `pollster` to be optional. Introduced new example paths for `sync` and added the `cocoa` and `objc` dependencies for macOS support.

Refactoring and new functionality:

* [`src/lib.rs`](diffhunk://#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R18-R26): Moved the `Mode` enum to a new `mode` module and updated the `detect` and `subscribe` functions to support both synchronous and asynchronous operations.
* [`src/mode.rs`](diffhunk://#diff-6c8a4892f149fb2549309424e7df5b76836ffeb4c8122a43fb25c312d9ac7f2eR1-R21): Created a new module for the `Mode` enum and its associated methods.
* [`src/platforms/*`](diffhunk://#diff-97a8fb02a23bc48f2010075057726c3669fa845f907fac7a8823cd398c46eb09R3-R11): Refactored platform-specific detection and subscription functions to support both synchronous and asynchronous operations. This includes changes in macOS, freedesktop, websys, and windows modules.

Example updates:

* [`examples/async.rs`](diffhunk://#diff-2a00a0a738adbdf3a604268c41784a47b619e63402725e2d5a7a6386b6a9e691R1-R19): Added a new asynchronous example demonstrating the use of the `detect` and `subscribe` functions.
* [`examples/sync.rs`](diffhunk://#diff-a25bf7d1ade078ace7013e8ecb0affd78ebee938b41ac399328efc54bcd8355bR1-R16): Added a new synchronous example demonstrating the use of the `detect` and `subscribe` functions.
* Removed outdated examples `detect.rs` and `notify.rs` as they were replaced by the new `async.rs` and `sync.rs` examples.

Fixes:
* Fixes #47 and #29, which caused macOS to detect the initial mode over and over again.
* Should also address #45 and #38

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -18,23 +26,14 @@ tokio = { version = "1.23.0", features = ["full"] }
...
-    "NSObject",
-    "NSString",
-] }
-objc2-app-kit = { version = "0.2.0", features = [
-    "NSAppearance",
-    "NSApplication",
-    "NSResponder",
-] }
+cocoa = "0.26.0"
+objc = "0.2.7"
```
[the remark this answers is not in the document]
Comment (edfloreshz) [reply Cargo.toml:36]: Oh i didn’t realize it, I’ll switch back

[the remark this answers is not in the document]
Comment (edfloreshz) [reply Cargo.toml]: Should be resolved.

**Comment by edfloreshz** [comment]
I'll go ahead and merge this.

File: src/platforms/macos/mod.rs [hunk 2]
```diff
@@ -1,2 +1,29 @@
...
+                // Compare with "Dark"
+                let dark_str = NSString::from_str("Dark");
+                let is_dark: bool = msg_send![style, isEqualToString:&*dark_str];
+                is_dark
+            } else {
+                false
+            }
+        } else {
+            false
+        }
```
Comment (madsmtm) [inline src/platforms/macos/mod.rs:27]: Tip: You can avoid the brittle `msg_send!` here by importing `NSUserDefaults` from `objc2-foundation` (might have to modify `Cargo.toml` to include the `"NSUserDefaults"` Cargo feature first).

**Comment by edfloreshz** [comment]
We could instead provide only the sync version of `detect` using `pollster` to handle `ashpd` and leave `subscribe` as async.

**Pull request merged**